mujoco_py/mjpid.pyx

```diff
@@ -24,77 +81,191 @@ cdef enum USER_DEFINED_ACTUATOR_PARAMS:
+            previous_errors=PIDErrors(
+                error=0.0,
+                derivative_error=0.0,
+                integral_error=d.userdata[id * NUM_USER_DATA_PER_ACT + IDX_CAS_INTEGRAL_ERROR],
+            )))
 
-    if fabs(error) < error_deadband:
-        error = 0.0
+        # Save errors
+        d.userdata[id * NUM_USER_DATA_PER_ACT + IDX_CAS_LAST_ERROR] = pos_output.errors.error
+        d.userdata[id * NUM_USER_DATA_PER_ACT + IDX_CAS_INTEGRAL_ERROR] = pos_output.errors.integral_error
+        des_vel = Kp_cas * (pos_output.errors.error + pos_output.errors.integral_error)
+    else:
+        # If P gain on position loop is zero, only use the velocity controller
+        des_vel = d.ctrl[id]
 
-    integral_error = d.userdata[id * NUM_USER_DATA_PER_ACT + IDX_INTEGRAL_ERROR]
-    integral_error += error * dt_in_sec
-    integral_error = fmax(-integral_max_clamp, fmin(integral_max_clamp, integral_error))
+    # Clamp max angular velocity
+    max_qvel = m.actuator_gainprm[id * NGAIN + IDX_CAS_MAX_VEL]
+    des_vel = fmax(-max_qvel, fmin(max_qvel, des_vel))
 
-    last_error = d.userdata[id * NUM_USER_DATA_PER_ACT + IDX_LAST_ERROR]
-    cdef double derivative_error = (error - last_error) / dt_in_sec
+    # Apply Exponential Moving Average smoothing to the velocity setpoint
+    ctrl_ema_vel = d.userdata[id * NUM_USER_DATA_PER_ACT + IDX_CAS_STORED_EMA_SMOOTH_V]
+    smoothing_factor = m.actuator_gainprm[id * NGAIN + IDX_CAS_EMA_SMOOTH_V]
 
-    derivative_error_last = d.userdata[id * NUM_USER_DATA_PER_ACT + IDX_DERIVATIVE_ERROR_LAST]
+    smoothed_vel_setpoint = (smoothing_factor * ctrl_ema_vel) + (1 - smoothing_factor) * des_vel
 
-    derivative_error = (1.0 - derivative_gain_smoothing) * derivative_error_last + \
-        derivative_gain_smoothing * derivative_error
+    d.userdata[id * NUM_USER_DATA_PER_ACT + IDX_CAS_STORED_EMA_SMOOTH_V] = smoothed_vel_setpoint
 
-    cdef double integral_error_term = 0.0
-    if integral_time_const != 0:
-        integral_error_term = integral_error / integral_time_const
+    vel_output = _pid(parameters=PIDParameters(
+        dt_seconds=dt_in_sec,
+        setpoint=smoothed_vel_setpoint,
+        feedback=d.actuator_velocity[id],
+        Kp=m.actuator_gainprm[id * NGAIN + IDX_CAS_PROPORTIONAL_GAIN_V],
+        error_deadband=0.0,
+        integral_max_clamp=m.actuator_gainprm[id * NGAIN + IDX_CAS_INTEGRAL_MAX_CLAMP_V],
+        integral_time_const=m.actuator_gainprm[id * NGAIN + IDX_CAS_INTEGRAL_TIME_CONSTANT_V],
+        derivative_gain_smoothing=0.0,
+        derivative_time_const=0.0,
+        previous_errors=PIDErrors(
+            error=0.0,
+            derivative_error=0.0,
+            integral_error=d.userdata[id * NUM_USER_DATA_PER_ACT + IDX_CAS_INTEGRAL_ERROR_V],
+        )))
 
-    cdef double derivative_error_term = derivative_error * derivate_time_const
+    # Save errors
+    d.userdata[id * NUM_USER_DATA_PER_ACT + IDX_CAS_LAST_ERROR_V] = vel_output.errors.error
```

Comment:
it seems this is not used?

Reply:
that's right. we don't have to use it since there's no D term. will rm.